test/test_functional_optim.py

```diff
@@ -17,6 +19,49 @@ def __init__(self):
     def forward(self, t1):
         return self.lin2(F.relu(self.lin1(t1)))
 
+# dummy class to showcase custom optimizer registration with functional wrapper
+class MyDummyFnOptimizer(object):
+    def __init__(
+        self,
+        params: List[Tensor],
+        lr: float = 1e-3,
+        betas: Tuple[float, float] = (0.9, 0.999),
+        eps: float = 1e-6,
+        weight_decay: float = 0.0,
+        _allow_empty_param_list: bool = False,
+    ):
+
+        if not 0.0 <= lr:
+            raise ValueError("Invalid learning rate: {}".format(lr))
+        if not 0.0 <= eps:
+            raise ValueError("Invalid epsilon value: {}".format(eps))
+        if not 0.0 <= betas[0] < 1.0:
+            raise ValueError("Invalid beta parameter at index 0: {}".format(betas[0]))
+        if not 0.0 <= betas[1] < 1.0:
+            raise ValueError("Invalid beta parameter at index 1: {}".format(betas[1]))
+        if not 0.0 < weight_decay:
+            raise ValueError("Invalid weight_decay value: {}".format(weight_decay))
+
+        self.defaults = {
+            "lr": lr,
+            "eps": eps,
+            "beta1": betas[0],
+            "beta2": betas[1],
+            "weight_decay": weight_decay,
+        }
+
+        if len(params) == 0 and not _allow_empty_param_list:
+            raise ValueError("optimizer got an empty parameter list")
+
+    def step_param(self, param: Tensor, grad: Optional[Tensor]):
+        # call the custom optimizer step_param implementation
+        with torch.no_grad():
+            raise RuntimeError("MyDummyFnOptimizer does not support step_param() as of now")
```

Comment:
"as of now" -> is there intention to modify this `step_param()` implementation for this `MyDummyFnOptimizer` in the future? Is it that you will add a unit test later that registers the functional optimizer and then actually uses it, at which point you will add the implementation for `step_param()`?

(same applies for `step()`)

Reply:
This is just to show case how the registration mechanic works. If we plan to upstream custom optimizers, then this can be overridden but I don't think its PT specific